app/controllers/user_badges_controller.rb

```diff
@@ -116,4 +121,14 @@ def can_assign_badge_to_user?(user)
   def ensure_badges_enabled
     raise Discourse::NotFound unless SiteSetting.enable_badges?
   end
+
+  def is_badge_reason_valid?(reason)
+    uri = URI.parse(reason)
+
+    if uri && uri.scheme && uri.host
```

Comment:
The whole `if`/`else` can be replace by this one-liner ;)

```suggestion
uri && uri.scheme && uri.host && Discourse.base_url == "#{uri.scheme}://#{uri.host}"
```

Reply:
Also, instead of checking that's the link is on the same host, we should check it's a working link to a post or a topic.

Reply:
Yes, we _almost_ do that in `TopicLink`. Here are the relevant lines

https://github.com/discourse/discourse/blob/851ef14096f0b6f3c6726151099483907f2c0805/app/models/topic_link.rb#L142-L158